torch/onnx/utils.py

```diff
@@ -1660,6 +1627,115 @@ def _export(
+    if export_type == _exporter_states.ExportTypes.PROTOBUF_FILE:
+        assert len(export_map) == 0
+        with torch.serialization._open_file_like(f, "wb") as opened_file:
+            opened_file.write(model_bytes)
+    elif export_type in [
+        _exporter_states.ExportTypes.ZIP_ARCHIVE,
+        _exporter_states.ExportTypes.COMPRESSED_ZIP_ARCHIVE,
+    ]:
+        compression = (
+            zipfile.ZIP_DEFLATED
+            if export_type == _exporter_states.ExportTypes.COMPRESSED_ZIP_ARCHIVE
+            else zipfile.ZIP_STORED
+        )
+        with zipfile.ZipFile(f, "w", compression=compression) as z:
+            z.writestr(_constants.ONNX_ARCHIVE_MODEL_PROTO_NAME, model_bytes)
+            for k, v in export_map.items():
+                z.writestr(k, v)
+    elif export_type == _exporter_states.ExportTypes.DIRECTORY:
+        if os.path.exists(f):
+            assert os.path.isdir(f)
+        else:
+            os.makedirs(f)
+
+        model_proto_file = os.path.join(f, _constants.ONNX_ARCHIVE_MODEL_PROTO_NAME)
+        with torch.serialization._open_file_like(model_proto_file, "wb") as opened_file:
+            opened_file.write(model_bytes)
+
+        for k, v in export_map.items():
+            weight_proto_file = os.path.join(f, k)
+            with torch.serialization._open_file_like(
+                weight_proto_file, "wb"
+            ) as opened_file:
+                opened_file.write(v)
+    else:
+        raise RuntimeError("Unknown export type")
+
+
+@_beartype.beartype
+def _add_onnxscript_fn(
+    model_bytes: bytes,
+    custom_opsets: Mapping[str, int],
+    val_use_external_data_format: bool,
+    model_file_location: str,
+) -> bytes:
+    """Insert model-included custom onnx-script function into ModelProto"""
+
+    # TODO(titaiwang): remove this when onnx becomes dependency
+    try:
+        import onnx
+    except ImportError:
+        raise errors.OnnxExporterError("Module onnx is not installed!")
+
+    # For > 2GB model, we need to load from file
+    # TODO(titaiwang): Does > 2GB model exists in model bytes?
+    # because in _export_onnx, the tensors should be saved separately
+    # from model proto already
+    if val_use_external_data_format:
+        # save the model bytes into onnx before using onnx.load
+        with torch.serialization._open_file_like(
+            model_file_location, "wb"
+        ) as opened_file:
+            opened_file.write(model_bytes)
+        model_proto = onnx.load(model_file_location, load_external_data=False)
+    else:
+        model_proto = onnx.load_from_string(model_bytes)
+    # Iterate graph nodes to insert only the included custom
+    # function_proto into model_proto
+    # TODO(titaiwang): Currently, onnxscript doesn't support
+    # ONNXFunction calling other ONNXFunction scenario, neither does it here
+    onnx_function_list = list()
+    included_node_func = set()
+    for node in model_proto.graph.node:
```

Comment:
Need to recursively check subgraph in nodes ... for operators like `onnx::Loop`, or `onnx::If` that has attribute of type `GraphProto`.

Reply:
Ah....

Reply:
Added a recursive function to get graphs from nodes, and a specific loop test in the next PR. LMK if you know what else we can do to make it even robust.